core/drivers/crypto/hisilicon/hpre_rsa.c

```diff
@@ -0,0 +1,1076 @@
+	}
+
+	for (lp_len = lhash_len; lp_len < db_len; lp_len++) {
+		if (db[lp_len] != 0)
+			break;
+	}
+
+	if (lp_len == db_len) {
+		EMSG("Fail to find fixed 01 in db");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	msg_len = db_len - lp_len - 1;
+	if (msg_len > rsa_data->message.length) {
+		EMSG("Message space is not enough");
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	*out_len = msg_len;
+	memcpy(out, db + lp_len + 1, msg_len);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rsa_oaep_decode(struct drvcrypt_rsa_ed *rsa_data,
+				  uint8_t *out, size_t *out_len)
+{
+	size_t lhash_len = rsa_data->digest_size;
+	uint8_t seed[OAEP_MAX_HASH_LEN] = { };
+	uint8_t db[OAEP_MAX_DB_LEN] = { };
+	uint8_t *mask_db = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	/* oaep format 00 || maskedseed || maskeddb */
+	mask_db = rsa_data->message.data + lhash_len + 1;
+	ret = rsa_oaep_get_seed(rsa_data, mask_db, seed);
+	if (ret)
+		return ret;
+
+	ret = rsa_oaep_get_db(rsa_data, mask_db, seed, db);
+	if (ret)
+		return ret;
+
+	return rsa_oaep_get_msg(rsa_data, db, out, out_len);
+}
+
+static TEE_Result rsa_oaep_decrypt(struct drvcrypt_rsa_ed *rsa_data)
+{
+	size_t n_bytes = rsa_data->key.n_size;
+	struct drvcrypt_rsa_ed rsa_dec_info = *rsa_data;
+	TEE_Result ret = TEE_SUCCESS;
+
+	/* Alloc oaep encode message data buf */
+	rsa_dec_info.message.data = malloc(n_bytes);
+	if (!rsa_dec_info.message.data) {
+		EMSG("Fail to alloc message data buf");
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	rsa_dec_info.message.length = n_bytes;
+	ret = rsa_nopad_decrypt(&rsa_dec_info);
+	if (ret)
+		goto free_data;
+
+	ret = rsa_oaep_decode(&rsa_dec_info, rsa_data->message.data,
+			      &rsa_data->message.length);
+	if (ret)
+		EMSG("Fail to get oaep decode message data");
+
+free_data:
+	free(rsa_dec_info.message.data);
+
+	return ret;
+}
+
+static TEE_Result hpre_rsa_decrypt(struct drvcrypt_rsa_ed *rsa_data)
+{
+	if (!rsa_data) {
+		EMSG("Invalid rsa decrypt input parameter");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	switch (rsa_data->rsa_id) {
+	case DRVCRYPT_RSA_NOPAD:
+	case DRVCRYPT_RSASSA_PKCS_V1_5:
+	case DRVCRYPT_RSASSA_PSS:
+		return rsa_nopad_decrypt(rsa_data);
+	case DRVCRYPT_RSA_PKCS_V1_5:
+		return rsa_pkcs_decrypt(rsa_data);
+	case DRVCRYPT_RSA_OAEP:
+		return rsa_oaep_decrypt(rsa_data);
+	default:
+		EMSG("Invalid rsa id");
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+}
+
+static const struct drvcrypt_rsa driver_rsa = {
+	.alloc_keypair = sw_crypto_acipher_alloc_rsa_keypair,
+	.alloc_publickey = hpre_rsa_allocate_publickey,
+	.free_publickey = hpre_rsa_free_publickey,
+	.free_keypair = sw_crypto_acipher_free_rsa_keypair,
+	.gen_keypair = sw_crypto_acipher_gen_rsa_key,
+	.encrypt = hpre_rsa_encrypt,
+	.decrypt = hpre_rsa_decrypt,
+	.optional = {
+		/*
+		 * If ssa_sign or verify is NULL, the framework will fill
+		 * data format directly by soft calculation. Then call api
+		 * encrypt or decrypt.
+		 */
+		.ssa_sign = NULL,
+		.ssa_verify = NULL,
+	},
```

Comment:
These explicit NULL inits are not required. But maybe you prefer to have them for the reader information.

Reply:
yes